croncheck.php

```diff
@@ -219,38 +219,31 @@
 
 $delay = '';
 $maxdelay = 0;
-$tasks = core\task\manager::get_all_scheduled_tasks();
-foreach ($tasks as $task) {
-    if ($task->get_disabled()) {
-        continue;
-    }
-    $faildelay = $task->get_fail_delay();
-    if ($faildelay == 0) {
-        continue;
-    }
-    if ($faildelay > $maxdelay) {
-        $maxdelay = $faildelay;
-    }
-    $delay .= "SCHEDULED TASK: " . get_class($task) . ' (' .$task->get_name() . ") Delay: $faildelay\n";
-}
 
-$records = $DB->get_records_sql('SELECT * from {task_adhoc} WHERE faildelay > 0');
-foreach ($records as $record) {
-    $task = \core\task\manager::adhoc_task_from_record($record);
-    if (!$task) {
-        continue;
-    }
+// Instead of using task API here, we read directly from the database.
+// This stops errors originating from broken tasks.
+$scheduledtasks = $DB->get_records_sql("SELECT * FROM {task_scheduled} WHERE faildelay > 0 AND disabled = 0");
+foreach ($scheduledtasks as $task) {
+    $delay .= "SCHEDULED TASK: {$task->classname} Delay: {$task->faildelay}\n";
+}
 
-    $faildelay = $task->get_fail_delay();
-    if ($faildelay == 0) {
-        continue;
-    }
-    if ($faildelay > $maxdelay) {
-        $maxdelay = $faildelay;
-    }
-    $delay .= "ADHOC TASK: " .get_class($task) . " Delay: $faildelay\n";
+// Instead of using task API here, we read directly from the database.
+// This stops errors originating from broken tasks, and allows the DB to de-duplicate them.
+$adhoctasks = $DB->get_records_sql("  SELECT classname, COUNT(*) count, MAX(faildelay) faildelay
```

Comment:
this all looks great, the only nit pick is adding line breaks to the sql after each comma

possibly you should also sort the tasks by cumulative faildelay

Reply:
> this all looks great, the only nit pick is adding line breaks to the sql after each comma

Do you mean:
```
SELECT classname,
COUNT(*) count,
MAX(faildelay) faildelay
...
```
?

> possibly you should also sort the tasks by cumulative faildelay

I have added this:
```
WARNING: Moodle task faildelay > 60 mins
ADHOC TASK: \tool_dataprivacy\task\process_data_request_task Delay: 1000  (13 duplicates!!!)
ADHOC TASK: \tool_dataprivacy\task\initiate_data_request_task Delay: 6001  (3 duplicates!!!)
 (Checked Oct 9 16:20:48)
```